ospfd/ospf_te.c

```diff
@@ -1588,6 +1603,1135 @@ void ospf_mpls_te_lsa_schedule(struct mpls_te_link *lp, enum lsa_opcode opcode)
+		  subnet->status == NEW ? "Add" : "Update", ls_pref->metric,
+		  &subnet->key);
+
+	/* Send corresponding Link State Message */
+	ospf_te_export(LS_MSG_TYPE_PREFIX, subnet);
+	subnet->status = SYNC;
+}
+
+static int ospf_te_parse_router_lsa(struct ls_ted *ted, struct ospf_lsa *lsa)
+{
+	struct router_lsa *rl;
+	enum node_type type;
+	struct ls_vertex *vertex;
+	struct ls_edge *edge;
+	struct ls_subnet *subnet;
+	struct listnode *node;
+	int len;
+
+	/* Sanity Check */
+	if (!ted || !lsa || !lsa->data)
+		return -1;
+
+	ote_debug("MPLS-TE (%s): Parse Router LSA[%pI4] from Router[%pI4]",
+		  __func__, &lsa->data->id, &lsa->data->adv_router);
+
+	/* Get vertex from LSA Advertise Router ID */
+	vertex = get_vertex(ted, lsa);
+
+	/* Set Node type information if it has changed */
+	rl = (struct router_lsa *)lsa->data;
+	if (IS_ROUTER_LSA_VIRTUAL(rl))
+		type = PSEUDO;
+	else if (IS_ROUTER_LSA_EXTERNAL(rl))
+		type = ASBR;
+	else if (IS_ROUTER_LSA_BORDER(rl))
+		type = ABR;
+	else
+		type = STANDARD;
+
+	if (vertex->status == NEW) {
+		vertex->node->type = type;
+		SET_FLAG(vertex->node->flags, LS_NODE_TYPE);
+	} else if (vertex->node->type != type) {
+		vertex->node->type = type;
+		vertex->status = UPDATE;
+	} else {
+		return 0;
+	}
+
+	ote_debug("  |- %s Vertex %pI4",
+		  vertex->status == NEW ? "Add" : "Update",
+		  &vertex->node->router_id);
+
+	/* Vertex has been created or updated: send corresponding LS Message */
+	ospf_te_export(LS_MSG_TYPE_NODE, vertex);
+	vertex->status = SYNC;
+
+	/* Mark outgoing Edge and Subnet as ORPHAN to detect deletion */
+	for (ALL_LIST_ELEMENTS_RO(vertex->outgoing_edges, node, edge))
+		edge->status = ORPHAN;
+
+	for (ALL_LIST_ELEMENTS_RO(vertex->prefixes, node, subnet))
+		subnet->status = ORPHAN;
+
+	/* Then, process Link Information */
+	len = ntohs(rl->header.length) - 4;
+	for (int i = 0; i < ntohs(rl->links) && len > 0; len -= 12, i++) {
+		struct prefix p;
+
+		switch (rl->link[i].type) {
+		case LSA_LINK_TYPE_POINTOPOINT:
+			ospf_te_update_link(ted, vertex, rl->link[i].link_data,
+					    ntohs(rl->link[i].metric));
+			/* Add corresponding subnet */
+			p.family = AF_INET;
+			p.prefixlen = 32;
+			p.u.prefix4 = rl->link[i].link_data;
+			ospf_te_update_subnet(ted, vertex, p,
+					      ntohs(rl->link[i].metric));
+			break;
+		case LSA_LINK_TYPE_STUB:
+			p.family = AF_INET;
+			p.prefixlen = ip_masklen(rl->link[i].link_data);
+			p.u.prefix4 = rl->link[i].link_id;
+			ospf_te_update_subnet(ted, vertex, p,
+					      ntohs(rl->link[i].metric));
+			break;
+		default:
+			break;
+		}
+	}
+	/* Clean remaining Orphan Edges or Subnets */
+	ls_vertex_clean(ted, vertex, zclient);
+
+	return 0;
+}
+
+static int ospf_te_delete_router_lsa(struct ls_ted *ted, struct ospf_lsa *lsa)
+{
+
+	struct ls_vertex *vertex;
+
+	/* Sanity Check */
+	if (!ted || !lsa || !lsa->data)
+		return -1;
+
+	/* Get vertex from LSA Advertise Router ID */
+	vertex = get_vertex(ted, lsa);
```

Comment:
The `get_vertex` creates the vertex if not exist and then the vertex is deleted, maybe the get_vertex fn could have a param like FORCE_CREATION |NOT_FORCE_CREATION in case it's really needed.

Reply:
Good catch. I will replace this call by a simple search like I done for edge or subnet deletion.
I'm preparing a major revision (bugs correction + check style + new topotest) that I'll send in an hour. Stay tune ;-)

Reply:
Done